rfcs/0000-structured-logging.md

```diff
@@ -0,0 +1,2189 @@
+
+struct MySource {
+    data: BTreeMap<String, serde_json::Value>,
+}
+
+impl Source for MySource {
+    fn visit<'kvs>(&'kvs self, visitor: &mut dyn source::Visitor<'kvs>) -> Result<(), source::Error> {
+        for (k, v) in self.data {
+            visitor.visit_pair(source::Key::new(k), source::Value::new(v))
+        }
+    }
+}
+```
+
+The `Key::new` method accepts any `T: Borrow<str>`. The `Value::new` accepts any `T: std::fmt::Debug + serde::Serialize`. Values that can't implement `Debug + Serialize` can still be visited using the `source::Value::any` method. This method lets us provide an inline function that will visit the value:
+
+```rust
+use log::kv::{
+    source::{self, Source},
+    value,
+};
+
+struct MySource {
+    data: BTreeMap<String, MyValue>,
+}
+
+impl Source for MySource {
+    fn visit<'kvs>(&'kvs self, visitor: &mut dyn source::Visitor<'kvs>) -> Result<(), source::Error> {
+        for (k, v) in self.data {
+            let key = source::Key::new(k);
+
+            // Let's assume `MyValue` doesn't implement `Serialize`
+            // Instead it implements `Display`.
+            let value = source::Value::any(v, |v, visitor| {
+                // Let's assume `MyValue` implements `Display`
+                visitor.visit_fmt(format_args!("{}", v))
+            });
+
+            visitor.visit_pair(key, value)
+    }
+}
+```
+
+A `Source` doesn't have to just contain key-value pairs directly like `BTreeMap<String, Value>` though. It could also act like an adapter, like we have for iterators in the standard library. As another example, the following `Source` doesn't store any key-value pairs of its own, it will sort and de-duplicate pairs read from another source by first reading them into a map before forwarding them on:
+
+```rust
+use log::kv::source::{self, Source, Visitor};
+
+pub struct SortRetainLast<KVS>(KVS);
+
+impl<KVS> Source for SortRetainLast<KVS>
+where
+    KVS: Source,
+{
+    fn visit<'kvs>(&'kvs self, visitor: &mut dyn source::Visitor<'kvs>) -> Result<(), source::Error> {
+        // `Seen` is a visitor that will capture key-value pairs
+        // in a `BTreeMap`. We use it internally to sort and de-duplicate
+        // the key-value pairs that `SortRetainLast` is wrapping.
+        struct Seen<'kvs>(BTreeMap<source::Key<'kvs>, source::Value<'kvs>>);
+
+        impl<'kvs> Visitor<'kvs> for Seen<'kvs> {
+            fn visit_pair<'vis>(&'vis mut self, k: source::Key<'kvs>, v: source::Value<'kvs>) -> Result<(), source::Error> {
+                self.0.insert(k, v);
+
+                Ok(())
+            }
+        }
+
+        // Visit the inner source and collect its key-value pairs into `seen`
+        let mut seen = Seen(BTreeMap::new());
+        self.0.visit(&mut seen)?;
+
+        // Iterate through the seen key-value pairs in order
+        // and pass them to the `visitor`.
+        for (k, v) in seen.0 {
+            visitor.visit_pair(k, v)?;
+        }
+
+        Ok(())
+    }
+}
+```
```

Comment:
All the code examples seem somewhat repetitive, any possibility of providing a derive macro? Like serde does.

Reply:
Ah these examples are all just variations of the same thing; implementing the `Source` trait where the underlying data is in various different shapes. I'm not sure a macro would be worthwhile for deriving `Source`, because the way you'd surface key-value pairs is really dependent on what the underlying source is.